engines/supernova/rooms.cpp

```diff
@@ -0,0 +1,3241 @@
+		_objectState[i]._direction = in->readByte();
+	}
+
+	_seen = in->readByte();
+
+	return !in->err();
+}
+
+Intro::Intro(SupernovaEngine *vm, GameManager *gm) {
+	_vm = vm;
+	_gm = gm;
+
+	_fileNumber = -1;
+	_id = INTRO;
+	_shown[0] = kShownFalse;
+
+	_objectState[0] =
+	Object(_id, kStringKeycard, kStringKeycardDescription, KEYCARD,
+		   TAKE | CARRIED | COMBINABLE, 255, 255, 0, NULLROOM, 0);
+	_objectState[1] =
+	Object(_id, kStringKnife, kStringKnifeDescription, KNIFE,
+		   TAKE | CARRIED | COMBINABLE, 255, 255, 0, NULLROOM, 0);
+	_objectState[2] =
+	Object(_id, kStringWatch, kStringDefaultDescription, WATCH,
+		   TAKE | COMBINABLE | CARRIED, 255, 255, 8, NULLROOM, 0);
+	_objectState[3] =
+	Object(_id, kStringDiscman, kStringDiscmanDescription, DISCMAN,
+		   TAKE | COMBINABLE, 255, 255, 0, NULLROOM, 0);
+	_objectState[4] =
+	Object(_id, kStringInventoryMoney, kStringDefaultDescription, MONEY,
+	       TAKE | COMBINABLE, 255, 255, 0);
+
+	_shouldExit = false;
+
+	introText =
+		_vm->getGameString(kStringIntro1) +
+		_vm->getGameString(kStringIntro2) +
+		_vm->getGameString(kStringIntro3) +
+		_vm->getGameString(kStringIntro4) +
+		_vm->getGameString(kStringIntro5) +
+		_vm->getGameString(kStringIntro6) +
+		_vm->getGameString(kStringIntro7) +
+		_vm->getGameString(kStringIntro8) +
+		_vm->getGameString(kStringIntro9) +
+		_vm->getGameString(kStringIntro10) +
+		_vm->getGameString(kStringIntro11) +
+		_vm->getGameString(kStringIntro12) +
+		_vm->getGameString(kStringIntro13);
+}
+
+void Intro::onEntrance() {
+	_gm->_guiEnabled = false;
+	_vm->_allowSaveGame = false;
+	_vm->_allowLoadGame = false;
+	titleScreen();
+	cutscene();
+	leaveCutscene();
+}
+
+void Intro::titleScreen() {
+	// Newspaper
+	CursorMan.showMouse(false);
+	_vm->_brightness = 0;
+	_vm->_menuBrightness = 0;
+	_vm->paletteBrightness();
+	_vm->setCurrentImage(1);
+	_vm->renderImage(0);
+	_vm->paletteFadeIn();
+	_gm->getInput();
+	_vm->paletteFadeOut();
+
+	// Title Screen
+	_vm->setCurrentImage(31);
+	_vm->renderImage(0);
+	_vm->paletteFadeIn();
+	_gm->wait2(1);
+	_vm->playSound(kAudioVoiceSupernova);
+	while (_vm->_mixer->isSoundHandleActive(_vm->_soundHandle))
+		_gm->wait2(1);
+	titleFadeIn();
+	_vm->renderText(kStringTitleVersion, 295, 190, kColorWhite44);
+	const Common::String& title1 = _vm->getGameString(kStringTitle1);
+	const Common::String& title2 = _vm->getGameString(kStringTitle2);
+	const Common::String& title3 = _vm->getGameString(kStringTitle3);
+	_vm->renderText(title1, 78 - _vm->textWidth(title1) / 2, 120, kColorLightBlue);
+	_vm->renderText(title2, 78 - _vm->textWidth(title2) / 2, 132, kColorWhite99);
+	_vm->renderText(title3, 78 - _vm->textWidth(title3) / 2, 142, kColorWhite99);
+	_gm->wait2(1);
+	CursorMan.showMouse(true);
+	_vm->playSoundMod(kMusicIntro);
+	_gm->getInput();
+	// TODO: render animated text
+	_vm->playSound(kAudioVoiceYeah);
+	while (_vm->_mixer->isSoundHandleActive(_vm->_soundHandle));
```

Comment:
Busy waiting is not good. Consider at least delaying and polling for events; and possibly updating the screen.

There are a few other instances of busy-waiting on `isSoundHandleActive` in this file that should be fixed.